packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+        {
+          code: BlockInputErrorCode.MISSING_TIME_COMPLETE,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getTimeComplete from BlockInput without a block and data"
+      );
+    }
+    return this.state.timeCompleteSec;
+  }
+
+  waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
+    if (!this.state.hasBlock) {
+      return withTimeout(() => this.blockPromise.promise, timeout, signal);
+    }
+    return Promise.resolve(this.state.block);
+  }
+  waitForAllData(timeout: number, signal?: AbortSignal): Promise<TData> {
+    return withTimeout(() => this.dataPromise.promise, timeout, signal);
+  }
+
+  async waitForBlockAndAllData(timeout: number, signal?: AbortSignal): Promise<this> {
+    if (!this.state.hasBlock || !this.state.hasAllData) {
+      await withTimeout(() => Promise.all([this.blockPromise.promise, this.dataPromise.promise]), timeout, signal);
+    }
+    return this;
+  }
+}
+
+// Pre-DA
+
+type BlockInputPreDataState = {
+  hasBlock: true;
+  hasAllData: true;
+  block: SignedBeaconBlock<ForkPreDeneb>;
+  source: SourceMeta;
+  timeCompleteSec: number;
+};
+
+/**
+ * Pre-DA, BlockInput only has a single state.
+ * - the block simply exists
+ */
+export class BlockInputPreData extends AbstractBlockInput<ForkPreDeneb, null> {
+  type = DAType.PreData as const;
+
+  state: BlockInputPreDataState;
+
+  constructor(init: BlockInputInit, state: BlockInputPreDataState) {
+    super(init);
+    this.state = state;
+  }
+
+  static createFromBlock(props: AddBlock & CreateBlockInputMeta): BlockInputPreData {
+    const init: BlockInputInit = {
+      daOutOfRange: props.daOutOfRange,
+      timeCreated: props.source.seenTimestampSec,
+      forkName: props.forkName,
+      slot: props.block.message.slot,
+      blockRootHex: props.blockRootHex,
+      parentRootHex: toHex(props.block.message.parentRoot),
+    };
+    const state: BlockInputPreDataState = {
+      hasBlock: true,
+      hasAllData: true,
+      block: props.block,
+      source: props.source,
+      timeCompleteSec: props.source.seenTimestampSec,
+    };
+    return new BlockInputPreData(init, state);
+  }
+
+  addBlock(_: AddBlock): void {
+    throw new BlockInputError(
+      {
+        code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+        blockRoot: this.blockRootHex,
+      },
+      "Cannot addBlock to BlockInputPreData"
+    );
+  }
+}
+
+// Blobs DA
+
+export type ForkBlobsDA = ForkName.deneb | ForkName.electra;
+
+type BlockInputBlobsState =
+  | {
+      hasBlock: true;
+      hasAllData: true;
+      versionHashes: VersionedHashes;
+      block: SignedBeaconBlock<ForkBlobsDA>;
+      source: SourceMeta;
+      timeCompleteSec: number;
+    }
+  | {
+      hasBlock: true;
+      hasAllData: false;
+      versionHashes: VersionedHashes;
+      block: SignedBeaconBlock<ForkBlobsDA>;
+      source: SourceMeta;
+    }
+  | {
+      hasBlock: false;
+      hasAllData: false;
+    };
```

Comment:
The commitments array is in the block header so the versioned hashes can be created on first blob input and the number of blobs will be known with either a block or the header so I suppose to be complete we should add a state for all the blobs came across but the block has not arrived yet (ie there is one blob for instance) and that way we do not wait and call for the data via reqresp

```suggestion
  | {
      hasBlock: true;
      hasAllData: false;
      versionHashes: VersionedHashes;
      block: SignedBeaconBlock<ForkBlobsDA>;
      source: SourceMeta;
    }
  | {
      hasBlock: false;
      hasAllData: true;
      versionHashes: VersionedHashes;
    }
  | {
      hasBlock: false;
      hasAllData: false;
    };
```

Reply:
The commitments array isn't in the block header, rather in the block body. (thats why the inclusion proof is added).

It's an interesting and subtle detail that column and blob sidecars are different in how they convey kzg commitments. Blob sidecars only include a single commitment, for the attached blob, whereas column sidecars include all commitments.

This means that all version hashes can be calculated after receiving a single column sidecar (and we will know how many blobs exist for that block).
But we will not be able to calculate all version hashes (or know how many blobs exist) after receiving a single blob sidecar.

Reply:
oh snap.  you are right!  good thing i didnt bet the farm on that one